storage/connect/tabrest.cpp

```diff
@@ -353,12 +186,140 @@ PTDB RESTDEF::GetTable(PGLOBAL g, MODE m)
   if (trace(515))
     htrc("REST GetTable mode=%d\n", m);
 
-  if (m != MODE_READ && m != MODE_READX && m != MODE_ANY) {
-    strcpy(g->Message, "REST tables are currently read only");
+  if (m != MODE_READ && m != MODE_READX && m != MODE_ANY)
+  {
+    snprintf(g->Message, sizeof(g->Message), "REST tables are currently read only");
     return NULL;
-  } // endif m
+  }
 
   return Tdp->GetTable(g, m); // Leave file type do the job
 } // end of GetTable
 
+
+/***********************************************************************/
+/*  curl_init: Initilize curl                                          */
+/***********************************************************************/
+int RESTDEF::curl_init(PGLOBAL g)
+{
+  CURLcode curl_res = curl_global_init(CURL_GLOBAL_ALL);
+  if (curl_res != CURLE_OK)
+  {
+    if (curl_res)
+    {
+      char msg[512];
+      snprintf(msg, 512, "unable to initialize curl library, "
+                         "curl returned this error code: %u "
+                         "with the following error message: %s",
+                         curl_res, curl_easy_strerror(curl_res));
+      snprintf(g->Message, sizeof(g->Message), "%s", msg);
+      return 1;
+    }
+  }
+  curl_inited = true;
+  return 0;
+}
+
+
+/***********************************************************************/
+/*  curl_deinit: Cleanup curl                                          */
+/***********************************************************************/
+void RESTDEF::curl_deinit()
+{
+  if (curl_inited)
+  {
+    curl_global_cleanup();
+    curl_inited = false;
+  }
+}
+
+
+/***********************************************************************/
+/*  WriteMemoryCallback: Curl callback function                        */
+/***********************************************************************/
+static size_t WriteMemoryCallback(void *contents,
```

Comment:
The default function writes to a file. Couldn't `CURLOPT_WRITEDATA` be used to set the `FILE*` pointer?

Reply:
I have tried this as the first approach in first testing, but the data returned in file were cut, since size cannot be known in advance.

Reply:
this sounds odd. Most web requests that curl would be programmed to do won't know the file size in advance.

Reply:
Worked with this:
```
diff --git a/storage/connect/tabrest.cpp b/storage/connect/tabrest.cpp
index 7413f6029b9..f976773f9a0 100644
--- a/storage/connect/tabrest.cpp
+++ b/storage/connect/tabrest.cpp
@@ -277,14 +277,11 @@ int RESTDEF::curl_run(PGLOBAL g)
   struct MemoryStruct chunk;
   chunk.memory = (char *)malloc(1);
   chunk.size = 0;
+  FILE *f= fopen(Fn, "wb");
 
   if ((curl_res= curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, curl_errbuf)) !=
         CURLE_OK ||
-      (curl_res= curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION,
-                                  WriteMemoryCallback)) !=
-        CURLE_OK ||
-      (curl_res= curl_easy_setopt(curl, CURLOPT_WRITEDATA,
-                                 (void *)&chunk)) !=
+      (curl_res= curl_easy_setopt(curl, CURLOPT_WRITEDATA, f)) !=
         CURLE_OK ||
       (curl_res = curl_easy_setopt(curl, CURLOPT_URL, buf)) != CURLE_OK ||
       (curl_res = curl_easy_perform(curl)) != CURLE_OK ||
@@ -303,8 +300,6 @@ int RESTDEF::curl_run(PGLOBAL g)
     }
   }
   curl_easy_cleanup(curl);
-  FILE *f= fopen(Fn, "wb");
-  fprintf(f, "%s", chunk.memory);
   fclose(f);
   free(chunk.memory);
   bool is_error = http_code < 200 || http_code >= 300;
```

* https://curl.se/libcurl/c/CURLOPT_WRITEDATA.html says libcurl DLL must have a WRITEFUNCTION.
* BB currently doesn't have/detect libcurl on amd-windows. MDBF-644
* Internally in curl, `fwrite` is used directly as the CURLOPT_WRITEFUNCTION. Unsure if that will directly work on Windows.